nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java

```diff
@@ -0,0 +1,585 @@
+                        modulePath = oldValue;
+                        logger.error(t.getLocalizedMessage(), t);
+                    }
+
+                } else if (SCRIPT_ENGINE.equals(descriptor)) {
+                    // The script engine has changed, so we need to set up a new instance for the selected
+                    // engine name
+                    scriptEngineName = newValue;
+                    setupEngine();
+                }
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    /**
+     * Creates a classloader to be used by the selected script engine and the provided script file. This
+     * classloader has the ExecuteScript's classloader as a parent (versus the current thread's context
+     * classloader) and also adds the specified module directory to the classpath. This enables scripts
+     * to use other scripts, modules, etc. without having to build them into the ExecuteScript NAR.
+     * If the parameter is null or empty, ExecuteScript's classloader is returned
+     *
+     * @param modulePath The path to a directory containing modules to be used by the script(s)
+     */
+    private ClassLoader createScriptEngineModuleClassLoader(String modulePath) {
+        URLClassLoader newModuleClassLoader = null;
+        if (StringUtils.isEmpty(modulePath)) {
+            return ExecuteScript.class.getClassLoader();
+        }
+        try {
+            newModuleClassLoader =
+                    new URLClassLoader(
+                            new URL[]{new File(modulePath).toURI().toURL()}, ExecuteScript.class.getClassLoader());
+        } catch (MalformedURLException mue) {
+            getLogger().error("Couldn't find modules directory at " + modulePath, mue);
+        }
+        return newModuleClassLoader;
+    }
+
+    /**
+     * Evaluates the given script body (or file) using the current session, context, and flowfile. The script
+     * evaluation expects a FlowFile to be returned, in which case it will route the FlowFile to success. If a script
+     * error occurs, the original FlowFile will be routed to failure. If the script succeeds but does not return a
+     * FlowFile, the original FlowFile will be routed to no-flowfile
+     *
+     * @param context the current process context
+     * @param session the current process session
+     * @throws ProcessException if any error occurs during script evaluation
+     */
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        synchronized (isInitialized) {
+            if (!isInitialized.get()) {
+                createResources();
+            }
+        }
+        ProcessorLog log = getLogger();
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        String[] args = context.getProperty(SCRIPT_ARGS).evaluateAttributeExpressions().getValue().split(";");
```

Comment:
I'm updating ExecuteScript to not need an input flow file, plus we're letting the script handle any incoming flow files, so they can get at the attributes they want. Supporting EL here was just to add a little extra power, to use things like ${hostname} and such, and was intended to usually be the same for each invocation of the script.

I will look at ArgumentUtils, probably should've been using that the whole time but I didn't know it was there :) Thanks!